### PR TITLE
[WIP] API for endpoints for client

### LINKED ISSUE #4175: API endpoints for clients
This belongs to the OAuth Epic: #4004 - please read this for the big picture of what this issue is for :)

---

Requires #4176

The following API endpoints are needed to make BREAD operations available for clients. These methods are needed for implementing the client management screen (see #4177). The endpoints are very similar to our other implementations and should follow the best practice used there.

- `GET /ghost/v0.1/clients`
- `GET /ghost/v0.1/clients/<id>`
- `GET /ghost/v0.1/clients/slug/<slug>`
- `POST /ghost/v0.1/clients/`
- `DELETE /ghost/v0.1/clients/<id>`
- `PUT /ghost/v0.1/clients/<id>`

Expected response format:
```
{
    clients:[
        {...}
    ]
}
```

### PR DESCRIPTION
**Pull request by delgermurun**
closes #4175 

Just saving my progress to github. But if someone has time, early review will be very much appreciated.

Thanks.

### CODE REVIEW

**Comment by delgermurun** [comment]
@ErisDS @sebgie 

I've done base prototype kind of thing. As far as I know, what is left is:
* Tests
* Permissions #4176 
* Adding/editing trusted domains
* Restrict editing default clients (`ghost-admin` and `ghost-client`)

Firstly, I need some help/direction for last 2 things.

How should add/editing trusted domains work? Sending array with data (`trustedDomains: ['ghost.org']`)? Is there any code that works similar?

Thanks.

**Comment by ErisDS** [comment]
Going to ping @kevinansfield (as the ember guru) and @zinyando (as the person working on the frontend) on this.

The only thing that works similarly, is tags, but I'm not sure that's done in a good way. Essentially, if you're always saving an array of things, you have to figure out what is being removed and what is being added.

For tags the code to update tags gets called [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/post.js#L60) (which is the dodgiest part, I think), the function lives [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/post.js#L193) and then this calls out to a set of functions [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/base/utils.js#L8). The logic works, but is very complex, and needs a proper home most likely.

I'm never sure, if updating all tags with a post (and all trusted domains with a client) makes most sense to be done this way, or whether it makes more sense to make individual calls to create / remove a trusted domain on a nested resource e.g. `POST /clients/2/trusted_domains/` and `DELETE /clients/2/trusted_domains/4/`.

I remember we talked about this a bit when we redid the tags input, but I can't remember the outcome. I think it might have been that it was too tricky to implement fresh at the stage we were but maybe @kevinansfield remembers more about what the pros/cons were.

**Comment by kevinansfield** [comment]
TL/DR:  :+1: for sticking with how we currently deal with posts/tags

The way that posts/tags work at the moment is definitely workable on the front-end but I can imagine it can make it more complicated on the backend if that method isn't baked into the data handling layer.

Ember Data's default approach is to have a separate REST endpoint for each resource (e.g. `clients` and `trusted_domains`) and have the relationships populated by supplying ID lists and/or side-loading the related records. However this does make it a bit more complicated in certain situations as it means the main object and it's relationships are saved in separate async operations and therefore managing failures is tricky because you need to manually roll back persisted data - I believe this is why we went with the embedded relationships approach.

JSON API has some discussion on ["multiple operation" support](https://github.com/json-api/json-api/issues/795) that would alleviate the the issue of multiple async API requests but there's no official spec yet and Ember's `EmbeddedRecordsMixin` essentially covers that use-case anyway.

File: core/server/models/base/utils.js [hunk 1]
```diff
@@ -67,4 +68,51 @@ tagUpdate = {
...
+        return function () {
+            return DomainModel.destroy({trusted_domain: domain.trusted_domain, client_id: client.id}, options);
+        };
+    },
+
+    createDomain: function createDomain(DomainModel, client, domain, options) {
+        return function () {
+            //return DomainModel.add({trusted_domain: domain.trusted_domain, client_id: client.id}, options);
+
+            return client.trustedDomains().create({trusted_domain: domain.trusted_domain}, options);
```
Comment (delgermurun) [inline core/server/models/base/utils.js]: I am trying to create trusted domain, but doesn't work.

@ErisDS can you help me, please?

Comment (ErisDS) [inline core/server/models/base/utils.js]: I think you need to use the same pattern as here: https://github.com/TryGhost/Ghost/pull/6195/files#diff-8ca6b928bab45f8c504b120eba34344dR36

Comment (ErisDS) [inline core/server/models/base/utils.js]: Also, come chat in #dev on [slack](https://ghost.org/slack) :wink:

File: core/server/models/client.js [hunk 2]
```diff
@@ -1,14 +1,188 @@
...
+                ));
+            });
+        }
+    },
+
+    trustedDomains: function trusted_domains() {
         return this.hasMany('ClientTrustedDomain', 'client_id');
+    },
+
+    trusted_domains: function trusted_domains2() {
```
Comment (delgermurun) [inline core/server/models/client.js]: Ghost convert query params to lowercase. If I pass `?include=trustedDomains`, it doesn't work. So added this function.

Should I just replace `trustedDomains` with `trusted_domains`?

/cc @ErisDS

Comment (ErisDS) [inline core/server/models/client.js]: Yes this function should be snake-case rather than camel, see the post model for an example.

**Comment by ErisDS** [comment]
Hey @delgermurun I realise it's been a little while. Would you be able to pick this back up again?

No worries if not, I can pick up where you've left off.

**Comment by delgermurun** [comment]
@ErisDS I've been very busy nowadays. Please pick up. Sorry.

**Comment by acburdine** [comment]
@ErisDS Not to steal this one away from you, but would it be alright if I took a crack a finishing this one up? Just want to re-familiarize myself with the backend

**Comment by ErisDS** [comment]
Please do 👍

**Comment by acburdine** [comment]
Closing in favor of #6674
